src/actions.ts

```diff
@@ -333,6 +335,29 @@ export function getActions(instance: PtzOpticsInstance): CompanionActionDefiniti
 				void instance.sendCommand(PresetRecall, event.options)
 			},
 		},
+		[PtzOpticsActionId.RecallPresetFromVar]: {
+			name: 'Recall Preset (by number)',
```

Comment:
I hate to jerk you around like this.  😢  But:

By mostly coincidence I just found myself in a somewhat-similar situation: having an action whose effect is to send a command, and there's a bug with the command sent, and so I needed to test it.  Specifically, the user-specified parameters in the "Custom command" action weren't being filled right.

And it occurred to me: the way these slightly-complicated actions work is you have computations of some sort, and they lead up to a `void instance.sendCommand(...)`.

But `sendCommand` is merely the top end of the stack here.  Really, you want to verify that the conversion of command/options to bytes produces the right thing -- because if that works, `sendCommand` will too.  (And if you're really worried about `sendCommand` not sending those computed bytes, then you should be writing a `RunCameraInteractionTest` test -- and there are plenty of those.)

So the right way to formulate this sort of thing, is to move most of the body of this action into a standalone function: https://github.com/bitfocus/companion-module-ptzoptics-visca/commit/a269cbf3a4073448d922746b9fb17ebbff31fddb.  And then you test _that_: https://github.com/bitfocus/companion-module-ptzoptics-visca/commit/7dca219d235afd141c918f156201743ac0f22e32.

Amusingly, I happen to have created a mock context as part of this, that I just moved into its own standalone file.  https://github.com/bitfocus/companion-module-ptzoptics-visca/commit/d6a9396012510fc33233f61894a342d5cbbda271

So you should be able to move most of this into a standalone `export async function parsePresetVariableOption(option: string, context: CompanionActionContext): Promise<CompanionOptionValues>`.  And then you can create one ` MockContext` per test (don't create one and managing its state across tests: Jest supports running tests in isolation, and that sort of cross-test dependency breaks doing so), and in each test you can verify that `PresetRecall.toBytes(options)` has the expected value, as happens in one of those prior rev links.

Reply:
@jswalden I'm happy to split this out, one concern I have though is that when I asked in slack about whether I should use context or instance for parsing variables, the answer was to use the instance vs the context.  I'm not sure what the implications are, but would prefer to follow that guidance.  Happy to take your context mock and merge it with some of the other instance mock needs (including log).  Thoughts?